plugins/guests/suse/cap/change_host_name.rb

```diff
@@ -4,19 +4,46 @@ module Cap
       class ChangeHostName
         def self.change_host_name(machine, name)
           comm = machine.communicate
-
           basename = name.split(".", 2)[0]
           if !comm.test('test "$(hostnamectl --static status)" = "#{basename}"', sudo: false)
             comm.sudo <<-EOH.gsub(/^ {14}/, '')
-              hostnamectl set-hostname '#{basename}'
-
-              # Prepend ourselves to /etc/hosts
-              grep -w '#{name}' /etc/hosts || {
-                sed -i'' '1i 127.0.0.1\\t#{name}\\t#{basename}' /etc/hosts
-              }
+            hostnamectl set-hostname '#{basename}'
+            echo #{name} > /etc/HOSTNAME
             EOH
+            network_with_hostname = machine.config.vm.networks.map {|t, c| c if c[:hostname] }.compact[0]
+            if network_with_hostname
+              replace_host(comm, name, basename, network_with_hostname[:ip])
+            else
+              add_hostname_to_loopback(comm, name, basename)
+            end
           end
         end
+
+        def self.add_hostname_to_loopback(comm, name, basename)
+          # Add hostname to a loopback address on /etc/hosts if not already there
+          # Will insert name at the first free address of the form 127.0.X.1, up to
+          # the loop_bound
+          loop_bound = 5
+          comm.sudo <<-EOH.gsub(/^ {14}/, '')
+          grep -w '#{name}' /etc/hosts || {
+            for i in {1..#{loop_bound}}; do
+              grep -w "127.0.${i}.1" /etc/hosts || {
+                echo "127.0.${i}.1 #{name} #{basename}" >> /etc/hosts
+                break
+              }
+            done
```

Comment:
I don't know how I feel about this.... It seems to work consistently, but also its a little janky?